File: listviewadapter/app/src/main/java/group/learn/listviewadapter/views/customadapter/CustomAdapter.java
package group.learn.listviewadapter.views.customadapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;
import group.learn.listviewadapter.R;
import group.learn.listviewadapter.data.Music;

import java.util.List;
import java.util.Objects;

public class CustomAdapter <T> extends BaseAdapter {

    private final Context context;
    private final List<T> data;

    public CustomAdapter(Context context, List<T> data) {
        this.context = context;
        this.data = data;
    }

    @Override
    public int getCount() {
        return data.size();
    }

    @Override
    public Object getItem(int position) {
        return data.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;

        if(Objects.isNull(convertView)){
           convertView =  LayoutInflater.from(context)
                    .inflate(R.layout.custom_list_item, parent, false);

           holder = new ViewHolder();
           holder.title = convertView.findViewById(R.id.title);
           holder.details = convertView.findViewById(R.id.details);
           convertView.setTag(holder);
        }else{
            holder = (ViewHolder) convertView.getTag();
        }

        T currentItem = data.get(position);
        if(currentItem instanceof String){
            holder.title.setText((CharSequence) currentItem);
        }

        return convertView;
    }


    static class ViewHolder{
        TextView title ;
        TextView details;
    }
}

File: 16_custom_list_view/app/src/main/java/com/java/a16_custom_list_view/CustomAdapter.kt
package com.java.a16_custom_list_view

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView

class CustomAdapter(
    private val context: Context,
    countryArrayList: ArrayList<CountryModelClass>
) : ArrayAdapter<CountryModelClass>(context, R.layout.item_list_layout, countryArrayList) {

    private class ViewHolder {
        var countryName: TextView? = null
        var countWin: TextView? = null
        var countryFlag: ImageView? = null
    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        var convertView = view
        val viewHolder: ViewHolder

        if (convertView == null) {
            val inflater = LayoutInflater.from(context)
            convertView = inflater.inflate(R.layout.item_list_layout, parent, false)

            viewHolder = ViewHolder()
            viewHolder.countryName = convertView.findViewById(R.id.country_name)
            viewHolder.countWin = convertView.findViewById(R.id.countwin)
            viewHolder.countryFlag = convertView.findViewById(R.id.country_flag)

            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as ViewHolder
        }

        val dataModel: CountryModelClass? = getItem(position)

        viewHolder.countryName?.text = dataModel?.countryName
        viewHolder.countWin?.text = dataModel?.winCount.toString()
        viewHolder.countryFlag?.setImageResource(dataModel?.imageFlag ?: R.drawable.ic_launcher_background)

        return convertView!!
    }

}
